rclpy/rclpy/node.py

```diff
@@ -761,6 +763,25 @@ def _set_parameters_atomically(
 
         return result
 
+    def add_on_set_paramters_callback(
+        self,
+        callback: Callable[[List[Parameter]], SetParametersResult]
+    ) -> None:
+        """Add the callback to list."""
+        prevCallBack = self._parameters_callback
+        prevCallBack.insert(0, callback)
+        self._parameters_callback = prevCallBack
+
+    def remove_on_set_parameters_callback(
+        self,
+        callback: Callable[[List[Parameter]], SetParametersResult]
+    ) -> None:
+        """Remove callback from list."""
+        if callback in self._parameters_callback:
+            self._parameters_callback.remove(callback)
+        else:
+            print('Callback does not exist')
```

Comment:
I feel like this should probably throw an exception instead.

Reply:
At first I did the exception one but I was told to print error message instead

Reply:
Are you referring to [this comment](https://github.com/ros2/rclpy/pull/427#discussion_r336711782)? Because I believe @jacobperron was actually saying you shouldn't catch the exception, but rather let it propagate up the stack.

Reply:
Okay sir..will update soon.
@allenh1 but sir why is my PR failing with tests?

Reply:
It appears to be failing because you have a situation where you have a parameter callback set, but that callback was not added to a list.

I suspect `set_parameters_callback` should be storing the new callback as a list with one element instead of just a callback. Again, defer to the behavior of the C++ implementation.

Reply:
@allenh1 sir but I have made two functions that for pushing callbacks in the callback list
`def add_on_set_parameter_callback`
`def remove_on_parameter_set_callback`
Are the test cases modified to check the current changes for callback lists..just asking  :)

Reply:
> Are the test cases modified to check the current changes for callback lists

No, not unless you modify them. I think the fix here is really simple, though. You should just be able to change the `set_parameters_callback` function to store the passed callback as `[callback]`. I would suggest that you do that instead of modifying the tests.

Reply:
@allenh1 okay sir

Reply:
regarding the removing of callback,in c++ code exception is thrown when a callback is not found..should I also implement the exception?

Reply:
@allenh1 ..yes sir testcases are passing now for parameter callbacks...thank you so much more helping..

Reply:
> yes sir testcases are passing now for parameter callbacks...thank you so much more helping

@suab321321 anytime! Glad to know it worked.

> it is only showing a DCO authorization error..I dont know why..

You need to do a `git commit --signoff` with your commits. You can modify your latest commit with this by doing `git commit --amend --signoff` and then a `git push --force-with-lease`.

> regarding the removing of callback,in c++ code exception is thrown when a callback is not found..should I also implement the exception?

I would say so, but it's not my place to comment on how that should behave. Maybe @wjwwood or @dirk-thomas has an opinion on this?

> will this be merged sir?

I think you should add some new tests that would exercise the new code you are contributing, but, again, this is not my call to make.

Reply:
@allenh1 sir I did this..but its not happening I dont know why.Here is error message
![Screenshot from 2019-11-12 20-36-16](https://user-images.githubusercontent.com/52123334/68683110-3e7a8a80-058c-11ea-8034-ea3c4a824e66.png)

Reply:
and regarding merging should I mention that it @ivanpauno for merging?

Reply:
According to the output there:

> Commit sha: 3a941ab, Author: Dirk Thomas, Committer: AbhinavSingh; Expected "Dirk Thomas dirk-thomas@users.noreply.github.com", but got "Abhinav Singh singhabhinav9051571833@gmail.com".

So it looks like it is wanting @dirk-thomas to signoff the first commit.

Reply:
Please keep iterating on the patch. When it is ready to be merged we can resolve the DCO failure.

Reply:
@dirk-thomas sir is my current changes ready to be merged?